include/lsst/afw/image/PhotoCalib.h

```diff
@@ -488,6 +489,19 @@ class PhotoCalib : public table::io::PersistableFacade<PhotoCalib>, public table
                                   ndarray::Array<double, 2, 2> result) const;
 };
 
+/**
+ * Construct a PhotoCalib from FITS FLUXMAG0/FLUXMAG0ERR keywords.
+ *
+ * This provides backwards compatibility with the obsoleted Calib object that PhotoCalib replaced.
+ * It should not be used outside of reading old Exposures written before PhotoCalib existed.
+ *
+ * @param metadata FITS header metadata containing FLUXMAG0 and FLUXMAG0ERR keys.
+ * @param strip Strip FLUXMAG0 and FLUXMAG0ERR from `metadata`?
+ *
+ * @throws lsst::pex::exceptions::TypeError if the metadata does not contain FLUXMAG0.
```

Comment:
I think just returning `nullptr` might be a more natural way to express this condition.

In fact, it looks like the old `Calib` code handled this case by returning a valid `Calib` initialized with zeros.  I'm by no means convinced that was a good idea, but if we do change that behavior, we should be prepared for breakage in places we didn't expect (though it may be that you're already dealt with that in a part of this ticket I haven't gotten to yet).

Reply:
I took the idea of raising from SkyWcs:

``` * @throws lsst::pex::exceptions::TypeError if the metadata does not describe a celestial WCS.
```

Can we try to do the Calib->PhotoCalib switch with the raise in place and then see if it does break anything? I would guess that the only place it would cause a problem are very, very old files that may be broken/invalid in other ways.

Returning nullptr feels more dangerous to me, since calling code then has to check on what it got.

Reply:
Fine with me.

My reasoning was that we'd actually (like WCS, I think?) run this on all files, and then overwrite it if we found a `Calib` or `PhotoCalib` in the afw::table::io binary tables, and I didn't want to have exceptions thrown in the usual case.  But that's an assumption about what we'll do on a later ticket.

In any case, note, that what matters in most cases is whether `Exposure.getCalib()` returns a `nullptr` when you couldn't read one, and I think that will be the case regardless of whether you throw or return `nullptr` here.

Reply:
> My reasoning was that we'd actually (like WCS, I think?) run this on all files, and then overwrite it if we found a `Calib` or `PhotoCalib` in the afw::table::io binary tables...

I'm not sure I understand: I thought the purpose of this was to read old `Calibs`, but not modify the files.

> whether you throw or return `nullptr` here.

Throwing means the calling code is immediately informed of a problem. But yes, we can keep `Exposure.getCalib()/Exposure.getPhotoCalib()` returning nullptr/None if it couldn't be read, which I think is better than initializing to 0.

Reply:
I meant "overwrite" in the in-memory `Exposure`, but I just realized I was talking about logic that I changed a couple of months ago - so "overwrite" is definitely the wrong word to use (sorry about the confusion!), but I think my point (which is a very minor one!) remains.

To be more explicit, the current logic in `ExposureFitsReader` [first tries to extract a `SkyWcs` from the metadata, inside a try block](https://github.com/lsst/afw/blob/master/src/image/ExposureFitsReader.cc#L50).  If that method throws, the data member that holds the `wcs` remains null.  If `makeSkyWcs` had returned null directly, then we'd have avoided using exceptions for non-exceptional-case control flow, which is a good thing (and maybe be better able to use exceptions to signify the genuinely exceptional case where there appears to be a WCS in the header but it's corrupted).  And I think return null there would match how we normally treat exposure components in other contexts - we don't expect `Exposure.getWcs()` to throw if there is no `Wcs`; we expect it to return null/`None`.

Anyhow, regardless of whether or not that works, `ExposureFitsReader` will [later try to read a `SkyWcs` from the afw::table::io archive binary tables](https://github.com/lsst/afw/blob/master/src/image/ExposureFitsReader.cc#L214).  If that fails, it will fall back the one it already loaded from the metadata, which may still be null if the on-disk file didn't actually have a WCS.

But this is indeed a very minor point, and if we were worried about `PhotoCalib`-from-metadata being called *outside* `ExposureFitsReader` (I'm pretty sure its `SkyWcs` counterpart is), then preferring an exception to a null return would be a perfectly good reason to just ignore my comment here, and maintaining similarity with `SkyWcs` is probably a perfectly good reason as well.

To summarize, I suppose I'm saying that:
 - I think I'd prefer `makeSkyWcs` to return null on normal-failure, for similarity with `exposure.getWcs()` and to avoid using exceptions for normal control flow.
 - Given that `makeSkyWcs` does throw (and changing it here is definitely out of scope), it's entirely reasonable for the `PhotoCalib` equivalent to either throw to be more like it, or to return null, to be more like `exposure.getCalib()` and to avoid using exceptions for normal control flow.

Reply:
`PhotoCalib` will not be persisted in the metadata, so there won't be a "fall back to metadata" option: the metadata reader is only for backwards compatibility with persisted `Calib`s.

Given that, I think you're right that it is better to return `nullptr` here instead of raising: existing Exposures that don't have that metadata won't behave any more oddly than they already do.

That said, throwing here doesn't feel like "normal flow control" to me: not having the values you expected when reading a file should be an "exceptional" case, but the old Exposure persistence wasn't careful enough for us to guarantee that. `PhotoCalibFactory::read`` will implicitly raise some kind of key error if the required fields are missing. Having fields missing from your persisted file is different from `get()`ing an (initialized or not) attribute on a constructed object.